src/EditorFeatures/CSharpTest/EditAndContinue/RudeEditStatementTests.cs

```diff
@@ -905,6 +905,529 @@ public void MatchExceptionHandlers()
+";
+
+            var match = GetMethodMatches(src1, src2, kind: MethodKind.Regular);
+            var actual = ToMatchingPairs(match);
+
+            var expected = new MatchingPairs
+            {
+                { "foreach (var (a3, a4) in e) { }", "foreach (var (a3, a5, a4) in e) { }" },
+                { "a3", "a3" },
+                { "a4", "a4" },
+                { "{ }", "{ }" },
+                { "foreach ((var b4, var b5) in e) { }", "foreach ((var b6, var b4, var b5) in e) { }" },
+                { "b4", "b4" },
+                { "b5", "b5" },
+                { "{ }", "{ }" }
+            };
+
+            expected.AssertEqual(actual);
+        }
+
+        [Fact]
+        public void MatchForeachVariable_Delete()
+        {
+            var src1 = @"
+foreach (var (a11, a12, a13) in e) { A1(); }
+foreach ((var b7, var b8, var b9) in e) { A2(); }
+";
+
+            var src2 = @"
+foreach (var (a12, a13) in e1) { A1(); }
+foreach ((var b7, var b9) in e) { A2(); }
+";
+
+            var match = GetMethodMatches(src1, src2, kind: MethodKind.Regular);
+            var actual = ToMatchingPairs(match);
+
+            var expected = new MatchingPairs
+            {
+                { "foreach (var (a11, a12, a13) in e) { A1(); }", "foreach (var (a12, a13) in e1) { A1(); }" },
+                { "a12", "a12" },
+                { "a13", "a13" },
+                { "{ A1(); }", "{ A1(); }" },
+                { "A1();", "A1();" },
+                { "foreach ((var b7, var b8, var b9) in e) { A2(); }", "foreach ((var b7, var b9) in e) { A2(); }" },
+                { "b7", "b7" },
+                { "b9", "b9" },
+                { "{ A2(); }", "{ A2(); }" },
+                { "A2();", "A2();" }
+            };
+
+            expected.AssertEqual(actual);
+        }
+
+        [Fact]
+        public void MatchConstantPattern()
+        {
+            var src1 = @"
+if ((o is null) && (y == 7)) return 3;
+if (a is 7) return 5;
+";
+
+            var src2 = @"
+if ((o1 is null) && (y == 7)) return 3;
+if (a is 77) return 5;
+";
+
+            var match = GetMethodMatches(src1, src2, kind: MethodKind.Regular);
+            var actual = ToMatchingPairs(match);
+
+            var expected = new MatchingPairs {
+                { "if ((o is null) && (y == 7)) return 3;", "if ((o1 is null) && (y == 7)) return 3;" },
+                { "return 3;", "return 3;" },
+                { "if (a is 7) return 5;", "if (a is 77) return 5;" },
+                { "return 5;", "return 5;" }
+            };
+
+            expected.AssertEqual(actual);
+        }
+
+        [Fact]
+        public void MatchDeclarationPattern()
+        {
+            var src1 = @"
+if (!(o is int i) && (y == 7)) return;
+if (!(a is string s)) return;
+if (!(b is string t)) return;
+if (!(c is int j)) return;
+";
+
+            var src2 = @"
+if (!(o1 is int i) && (y == 7)) return;
+if (!(a is int s)) return;
+if (!(b is string t1)) return;
```

Comment:
Perhaps make it more interesting by reordering these. The match should not change if you do so.